Add `f16` and `f128` math functions

tgross35 opened the pull request:
This adds intrinsics and math functions for `f16` and `f128` floating point types. Support is quite limited and some things are broken so tests don't run on many platforms, but this provides a starting point.

try-job: aarch64-gnu

<!-- homu-ignore:start -->
r? ghost
@rustbot label +F-f16_and_f128
<!-- homu-ignore:end -->

tgross35 committed:
Add math intrinsics for `f16` and `f128`

These already exist in the compiler. Expose them in core so we can add
their library functions.

tgross35 committed:
Add math functions for `f16` and `f128`

This adds missing functions for math operations on the new float types.

Platform support is pretty spotty at this point, since even platforms
with generally good support can be missing math functions.
`std/build.rs` is updated to reflect this.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9693939312/job/26750493345) [(plain)](https://github.com/rust-lang/rust/commit/1f04794dc7c1558b8fe3e9822127f9e8e1c0220c/checks/26750493345/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling addr2line v0.22.0
thread 'rustc' panicked at /checkout/compiler/rustc_const_eval/src/interpret/cast.rs:190:29:
not implemented: f16_f128
stack backtrace:
   0:     0x7f21444f4e3f - <std[1e810906d2fadf8]::sys::backtrace::_print::DisplayBacktrace as core[839ef12c70fa2509]::fmt::Display>::fmt
   1:     0x7f21445513b0 - core[839ef12c70fa2509]::fmt::write
   2:     0x7f21444e9849 - <std[1e810906d2fadf8]::sys::pal::unix::stdio::Stderr as std[1e810906d2fadf8]::io::Write>::write_fmt
   3:     0x7f21444f4c0e - std[1e810906d2fadf8]::sys::backtrace::print
   4:     0x7f21444f78ea - std[1e810906d2fadf8]::panicking::default_hook::{closure#1}
   5:     0x7f21444f7574 - std[1e810906d2fadf8]::panicking::default_hook
   6:     0x7f214624e307 - <alloc[8501d074461e147f]::boxed::Box<rustc_driver_impl[f082d0f2c590f52a]::install_ice_hook::{closure#0}> as core[839ef12c70fa2509]::ops::function::Fn<(&dyn for<'a, 'b> core[839ef12c70fa2509]::ops::function::Fn<(&'a std[1e810906d2fadf8]::panic::PanicHookInfo<'b>,), Output = ()> + core[839ef12c70fa2509]::marker::Send + core[839ef12c70fa2509]::marker::Sync, &std[1e810906d2fadf8]::panic::PanicHookInfo)>>::call
   7:     0x7f21444f82d8 - std[1e810906d2fadf8]::panicking::rust_panic_with_hook
   8:     0x7f21444f7e63 - std[1e810906d2fadf8]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f21444f5409 - std[1e810906d2fadf8]::sys::backtrace::__rust_end_short_backtrace::<std[1e810906d2fadf8]::panicking::begin_panic_handler::{closure#0}, !>
  11:     0x7f214454d3a3 - core[839ef12c70fa2509]::panicking::panic_fmt
  11:     0x7f214454d3a3 - core[839ef12c70fa2509]::panicking::panic_fmt
  12:     0x7f21473c4780 - <rustc_const_eval[c98065791f7a8716]::interpret::eval_context::InterpCx<rustc_const_eval[c98065791f7a8716]::const_eval::dummy_machine::DummyMachine>>::float_to_float_or_int
  13:     0x7f21473fd8ee - <rustc_mir_transform[758d6a3f5e14248e]::gvn::VnState>::eval_to_const::{closure#0}
  14:     0x7f21473fa17a - <rustc_mir_transform[758d6a3f5e14248e]::gvn::VnState>::insert
  15:     0x7f2147402e75 - <rustc_mir_transform[758d6a3f5e14248e]::gvn::VnState>::simplify_rvalue::{closure#0}
  16:     0x7f2147401eca - <rustc_mir_transform[758d6a3f5e14248e]::gvn::VnState>::simplify_rvalue
  17:     0x7f21472d149b - <rustc_mir_transform[758d6a3f5e14248e]::ssa::SsaLocals>::for_each_assignment_mut::<rustc_mir_transform[758d6a3f5e14248e]::gvn::propagate_ssa::{closure#0}>
  18:     0x7f21473f830b - <rustc_mir_transform[758d6a3f5e14248e]::gvn::GVN as rustc_middle[181fecaaf64f9719]::mir::MirPass>::run_pass
  19:     0x7f2147301517 - rustc_mir_transform[758d6a3f5e14248e]::pass_manager::run_passes_inner
  20:     0x7f2147311862 - rustc_mir_transform[758d6a3f5e14248e]::optimized_mir
  21:     0x7f21484438d7 - rustc_query_impl[bd46294947b1e958]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bd46294947b1e958]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[181fecaaf64f9719]::query::erase::Erased<[u8; 8usize]>>
  22:     0x7f2148870009 - <rustc_query_impl[bd46294947b1e958]::query_impl::optimized_mir::dynamic_query::{closure#2} as core[839ef12c70fa2509]::ops::function::FnOnce<(rustc_middle[181fecaaf64f9719]::ty::context::TyCtxt, rustc_span[802050f372e962ba]::def_id::DefId)>>::call_once
  23:     0x7f21484bb511 - rustc_query_system[41a8177b433cd17]::query::plumbing::try_execute_query::<rustc_query_impl[bd46294947b1e958]::DynamicConfig<rustc_query_system[41a8177b433cd17]::query::caches::DefIdCache<rustc_middle[181fecaaf64f9719]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[bd46294947b1e958]::plumbing::QueryCtxt, false>
  24:     0x7f214869cec7 - rustc_query_impl[bd46294947b1e958]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7f2149a21fc0 - rustc_middle[181fecaaf64f9719]::query::plumbing::query_get_at::<rustc_query_system[41a8177b433cd17]::query::caches::DefIdCache<rustc_middle[181fecaaf64f9719]::query::erase::Erased<[u8; 8usize]>>>
  26:     0x7f2149a279cf - <rustc_metadata[fbc17f3b20edd792]::rmeta::encoder::EncodeContext>::encode_crate_root
  27:     0x7f2149a36f42 - rustc_metadata[fbc17f3b20edd792]::rmeta::encoder::encode_metadata
  28:     0x7f2149a5bb87 - rustc_metadata[fbc17f3b20edd792]::fs::encode_and_write_metadata
  29:     0x7f21464b36b6 - rustc_interface[71536429767fb805]::passes::start_codegen
  30:     0x7f214662d000 - <rustc_middle[181fecaaf64f9719]::ty::context::GlobalCtxt>::enter::<<rustc_interface[71536429767fb805]::queries::Queries>::codegen_and_build_linker::{closure#0}, core[839ef12c70fa2509]::result::Result<rustc_interface[71536429767fb805]::queries::Linker, rustc_span[802050f372e962ba]::ErrorGuaranteed>>
  31:     0x7f214668086e - <rustc_interface[71536429767fb805]::queries::Queries>::codegen_and_build_linker
  32:     0x7f21461e24a3 - <rustc_interface[71536429767fb805]::interface::Compiler>::enter::<rustc_driver_impl[f082d0f2c590f52a]::run_compiler::{closure#0}::{closure#1}, core[839ef12c70fa2509]::result::Result<core[839ef12c70fa2509]::option::Option<rustc_interface[71536429767fb805]::queries::Linker>, rustc_span[802050f372e962ba]::ErrorGuaranteed>>
  33:     0x7f214623a9a0 - rustc_span[802050f372e962ba]::create_session_globals_then::<core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>, rustc_interface[71536429767fb805]::util::run_in_thread_with_globals<rustc_interface[71536429767fb805]::util::run_in_thread_pool_with_globals<rustc_interface[71536429767fb805]::interface::run_compiler<core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>, rustc_driver_impl[f082d0f2c590f52a]::run_compiler::{closure#0}>::{closure#1}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
  34:     0x7f2146284c52 - std[1e810906d2fadf8]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[71536429767fb805]::util::run_in_thread_with_globals<rustc_interface[71536429767fb805]::util::run_in_thread_pool_with_globals<rustc_interface[71536429767fb805]::interface::run_compiler<core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>, rustc_driver_impl[f082d0f2c590f52a]::run_compiler::{closure#0}>::{closure#1}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>
  35:     0x7f21461d41ff - <<std[1e810906d2fadf8]::thread::Builder>::spawn_unchecked_<rustc_interface[71536429767fb805]::util::run_in_thread_with_globals<rustc_interface[71536429767fb805]::util::run_in_thread_pool_with_globals<rustc_interface[71536429767fb805]::interface::run_compiler<core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>, rustc_driver_impl[f082d0f2c590f52a]::run_compiler::{closure#0}>::{closure#1}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[839ef12c70fa2509]::result::Result<(), rustc_span[802050f372e962ba]::ErrorGuaranteed>>::{closure#2} as core[839ef12c70fa2509]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7f2144503aa4 - <std[1e810906d2fadf8]::sys::pal::unix::thread::Thread>::new::thread_start
  38:     0x7f214431b850 - <unknown>
  39:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

warning: the ICE couldn't be written to `/checkout/rustc-ice-2024-06-27T09_23_37-12375.txt`: Read-only file system (os error 30)
note: rustc 1.81.0-nightly (83b82d019 2024-06-27) running on x86_64-unknown-linux-gnu


note: compiler flags: --crate-type dylib --crate-type rlib -C prefer-dynamic -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -Z unstable-options -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C embed-bitcode=yes -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [optimized_mir] optimizing MIR for `f16::<impl at library/std/src/f16.rs:19:1: 19:9>::asinh`
error: could not compile `std` (lib)
Build completed unsuccessfully in 0:04:39
+ set -e
+ cat /tmp/toolstate/toolstates.json
+ cat /tmp/toolstate/toolstates.json
{"lld-wrapper":"test-fail"}
+ python3 ../x.py test --stage 2 check-tools
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
ERROR: Tool `book` was not recorded in tool state.
ERROR: Tool `nomicon` was not recorded in tool state.
ERROR: Tool `reference` was not recorded in tool state.
ERROR: Tool `rust-by-example` was not recorded in tool state.
ERROR: Tool `edition-guide` was not recorded in tool state.
ERROR: Tool `embedded-book` was not recorded in tool state.
  local time: Thu Jun 27 09:23:43 UTC 2024
  network time: Thu, 27 Jun 2024 09:23:43 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented:
@rustbot label +rla-silenced

tgross35 committed:
Specify the integer type of the `powi` LLVM intrinsic

Since LLVM <https://reviews.llvm.org/D99439> (4c7f820b2b20, "Update
@llvm.powi to handle different int sizes for the exponent"), the size of
the integer can be specified for the `powi` intrinsic. Make use of this
so it is more obvious that integer size is consistent across all float
types.

This feature is available since LLVM 13 (October 2021). Based on
bootstrap we currently support >= 17.0, so there should be no support
problems.

tgross35 commented:
Still getting a failure on f16 locally (aarch64) that I can't explain, `1.0f16.powi(1)` is returning a wrong result. This happens in the unit tests but doesn't seem to replicate when I copy them to a separate file, making this difficult to debug.

May as well see if this shows up in CI.

@bors try

bors commented:
:hourglass: Trying commit 0bca7c358be29fb03f1764f43f8ca700dfc01195 with merge 4ad38c5589324c480a08ee92cdf1028bed0d5690...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0bca7c358be29fb03f1764f43f8ca700dfc01195","merge_sha":"4ad38c5589324c480a08ee92cdf1028bed0d5690"} -->

tgross35 commented:
Figured out how to reproduce, the incorrect value is only returned if optimizations are on 💀 

@bors cancel

tgross35 commented:
@bors try-

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9904783637/job/27362925215) [(plain)](https://github.com/rust-lang-ci/rust/commit/4ad38c5589324c480a08ee92cdf1028bed0d5690/checks/27362925215/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
[RUSTC-TIMING] minmax_stability_issue_23687 test:true 0.987
error: format argument must be a string literal
   --> library/std/src/f128/tests.rs:543:40
    |
543 |     assert_eq!(pi.to_degrees(), 180.0, TOL_P2);
    |
help: you might be missing a string literal to format with
    |
    |
543 |     assert_eq!(pi.to_degrees(), 180.0, "{}", TOL_P2);

[RUSTC-TIMING] vec_deque_alloc_error test:true 0.444
[RUSTC-TIMING] seq_compare test:true 0.385
[RUSTC-TIMING] env test:true 1.400
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9904783637/job/27364980715)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9904783637/job/27364980715","builder_name":"checks-actions"} -->

tgross35 commented:
f16 issue on aarch: https://github.com/llvm/llvm-project/issues/98665

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 9dd7bb84cbd6b1b743f8fa087258da120b7f603f with merge 977d16ea2f47b72e115d1429e598a98d01384e13...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9dd7bb84cbd6b1b743f8fa087258da120b7f603f","merge_sha":"977d16ea2f47b72e115d1429e598a98d01384e13"} -->

tgross35 commented:
There are probably some rough edges that I need to smooth out and some tests that need tweaking, but this PR is somewhat large so I think somebody can start taking a look at it. f16 pass on my aarch machine, I can't test f128 locally so I'm going via try jobs.

Note that this is still going to fail the check CI job until nightly clippy becomes beta next week.

r? @dtolnay 

Since I know you have reviewed the math functions before. Feel free to reroll if you prefer.

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9911923947/job/27385631651) [(plain)](https://github.com/rust-lang-ci/rust/commit/977d16ea2f47b72e115d1429e598a98d01384e13/checks/27385631651/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #127020) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit ea38a04ac8466b0b4e55d3596bd4e22396c318f8 with merge eea39efc51d029ad4bcae43dbd5cf5d34216e3b7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ea38a04ac8466b0b4e55d3596bd4e22396c318f8","merge_sha":"eea39efc51d029ad4bcae43dbd5cf5d34216e3b7"} -->

tgross35 commented:
@bors try-

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9953112267/job/27495876446) [(plain)](https://github.com/rust-lang-ci/rust/commit/eea39efc51d029ad4bcae43dbd5cf5d34216e3b7/checks/27495876446/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 654257d516db27c6c78caa633c5863ae63b07fd6 with merge e2a4c6fef70695f68f9d5796aa18c835ed0e5cbf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"654257d516db27c6c78caa633c5863ae63b07fd6","merge_sha":"e2a4c6fef70695f68f9d5796aa18c835ed0e5cbf"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9954938443/job/27501664126) [(plain)](https://github.com/rust-lang-ci/rust/commit/e2a4c6fef70695f68f9d5796aa18c835ed0e5cbf/checks/27501664126/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling alloc v0.0.0 (/checkout/library/alloc)
[RUSTC-TIMING] eq_multidispatch test:true 0.219
[RUSTC-TIMING] log_knows_the_names_of_variants_in_std test:true 0.304
error: no rules expected the token `TOL_P2`
   --> library/std/src/f128/tests.rs:450:86
    |
450 |     assert_approx_eq!((-3.1f128).powf(2.0), 9.61000000000000055067062021407765193876 TOL_P2);
    |                                                                                     -^^^^^^ no rules expected this token in macro call
    |                                                                                     help: missing comma here
    |
   ::: library/std/src/macros.rs:378:1
    |
    |
378 | macro_rules! assert_approx_eq {
    | ----------------------------- when calling this macro
    |
note: while trying to match meta-variable `$b:expr`
    |
    |
379 |     ($a:expr, $b:expr) => {{ assert_approx_eq!($a, $b, 1.0e-6) }};

[RUSTC-TIMING] proc_macro test:true 2.380
[RUSTC-TIMING] volatile_fat_ptr test:true 0.214
[RUSTC-TIMING] create_dir_all_bare test:true 0.666
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9954938443/job/27503640380)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9954938443/job/27503640380","builder_name":"checks-actions"} -->

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 55fdeac2589573c89399704c3179fde5be1ba674 with merge fb5c0b7076e26935648a8aec525f517d5048579b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"55fdeac2589573c89399704c3179fde5be1ba674","merge_sha":"fb5c0b7076e26935648a8aec525f517d5048579b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9962286797/job/27527629061)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9962286797/job/27527629061","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9962286797/job/27525716222) [(plain)](https://github.com/rust-lang-ci/rust/commit/fb5c0b7076e26935648a8aec525f517d5048579b/checks/27525716222/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

tgross35 commented:
@bors try
@rustbot label +rla-silenced

bors commented:
:hourglass: Trying commit 858669faad764dfa9e3936db9bad023cd55a91ff with merge e3f08b398e6ce8cfeb1eafc874aa8ff86d83b955...
<!-- homu: {"type":"TryBuildStarted","head_sha":"858669faad764dfa9e3936db9bad023cd55a91ff","merge_sha":"e3f08b398e6ce8cfeb1eafc874aa8ff86d83b955"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9964587275/job/27533193777) [(plain)](https://github.com/rust-lang-ci/rust/commit/e3f08b398e6ce8cfeb1eafc874aa8ff86d83b955/checks/27533193777/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
[RUSTC-TIMING] builtin_clone test:true 0.307
[RUSTC-TIMING] env test:true 1.611
[RUSTC-TIMING] stdbenches test:true 2.666
[RUSTC-TIMING] test test:true 15.006
error: constant `TOL_N2` is never used
  --> library/std/src/f128/tests.rs:10:7
   |
10 | const TOL_N2: f128 = 1e-32;
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9964587275/job/27534654917)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9964587275/job/27534654917","builder_name":"checks-actions"} -->

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 0d5e03dfd9e21ceb9b3cd6ce357419f351a917b0 with merge a704ea08a642585f86e7b5021a2abd31733e3cbe...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0d5e03dfd9e21ceb9b3cd6ce357419f351a917b0","merge_sha":"a704ea08a642585f86e7b5021a2abd31733e3cbe"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9965241614/job/27535163810) [(plain)](https://github.com/rust-lang-ci/rust/commit/a704ea08a642585f86e7b5021a2abd31733e3cbe/checks/27535163810/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9965241614/job/27536351946)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9965241614/job/27536351946","builder_name":"checks-actions"} -->

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 1d270951953def6b3874cb50c8d86290d4674c8d with merge 82d8c3705b82c67b8152b503fe04431c6497811c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1d270951953def6b3874cb50c8d86290d4674c8d","merge_sha":"82d8c3705b82c67b8152b503fe04431c6497811c"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9971701235/job/27553356526) [(plain)](https://github.com/rust-lang-ci/rust/commit/82d8c3705b82c67b8152b503fe04431c6497811c/checks/27553356526/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9971701235/job/27555756520)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9971701235/job/27555756520","builder_name":"checks-actions"} -->

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 4824e0e25a555f46921f77a11d65ee2a485a70b3 with merge e6c896bb280e5b21c56d50f34a7c59097004d317...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4824e0e25a555f46921f77a11d65ee2a485a70b3","merge_sha":"e6c896bb280e5b21c56d50f34a7c59097004d317"} -->

tgross35 commented:
@rustbot ready

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9972818903/job/27556881953) [(plain)](https://github.com/rust-lang-ci/rust/commit/e6c896bb280e5b21c56d50f34a7c59097004d317/checks/27556881953/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
warning: profiler_builtins@0.0.0:       |         ^~~~~~~~~~~~~~~~~~~~~
error: unresolved link to `f16::min`
    --> library/core/src/intrinsics.rs:1891:11
     |
1891 |     /// [`f16::min`]
     |           ^^^^^^^^ the primitive type `f16` has no associated item named `min`
     = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: unresolved link to `f128::min`
error: unresolved link to `f128::min`
    --> library/core/src/intrinsics.rs:1927:11
     |
1927 |     /// [`f128::min`]
     |           ^^^^^^^^^ the primitive type `f128` has no associated item named `min`
error: unresolved link to `f16::max`
    --> library/core/src/intrinsics.rs:1940:11
     |
     |
1940 |     /// [`f16::max`]
     |           ^^^^^^^^ the primitive type `f16` has no associated item named `max`
error: unresolved link to `f128::max`
    --> library/core/src/intrinsics.rs:1976:11
     |
     |
1976 |     /// [`f128::max`]
     |           ^^^^^^^^^ the primitive type `f128` has no associated item named `max`
error: could not document `core`
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] core test:false 28.104


Command cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_aarch64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_CODEGEN_UNITS="1" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-std" CC_aarch64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="aarch64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="aarch64-unknown-linux-gnu" CFG_RELEASE_CHANNEL="nightly" CFLAGS_aarch64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC" CXXFLAGS_aarch64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC" CXX_aarch64_unknown_linux_gnu="sccache c++" LIBC_CHECK_CFG="1" RANLIB_aarch64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/aarch64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_FORCE_UNSTABLE="1" RUSTC_HOST_FLAGS="-Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_PRINT_STEP_TIMINGS="1" RUSTC_REAL="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/lib" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="-Csymbol-mangling-version=legacy -Zunstable-options --check-cfg=cfg(feature,values(any())) -Zunstable-options --check-cfg=cfg(bootstrap) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.81.0-nightly\t(e6c896bb2\t2024-07-17) -Zcrate-attr=doc(html_root_url=\"https://doc.rust-lang.org/nightly/\") -Zcrate-attr=warn(rust_2018_idioms) -Z unstable-options --resource-suffix 1.81.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md" RUSTDOC_LIBDIR="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" RUSTFLAGS="-Csymbol-mangling-version=legacy -Zunstable-options --check-cfg=cfg(feature,values(any())) -Zunstable-options --check-cfg=cfg(bootstrap) -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Zinline-mir-preserve-debug -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Cembed-bitcode=yes -Cforce-frame-pointers=yes -Zcrate-attr=doc(html_root_url=\"https://doc.rust-lang.org/nightly/\")" RUST_COMPILER_RT_ROOT="/checkout/src/llvm-project/compiler-rt" RUST_TEST_THREADS="8" TERM="xterm" WINAPI_NO_BUNDLED_LIBRARIES="1" __CARGO_DEFAULT_LIB_METADATA="nightlystd" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "doc" "--target" "aarch64-unknown-linux-gnu" "--release" "-Zbinary-dep-depinfo" "-j" "8" "--locked" "--color" "always" "--features" " panic-unwind backtrace profiler compiler-builtins-c" "--manifest-path" "/checkout/library/sysroot/Cargo.toml" "--no-deps" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-std/aarch64-unknown-linux-gnu/doc" "-Zskip-rustdoc-fingerprint" "-p" "alloc" "-p" "core" "-p" "panic_abort" "-p" "panic_unwind" "-p" "proc_macro" "-p" "profiler_builtins" "-p" "std" "-p" "test" "-p" "unwind" (failure_mode=Exit, stdout_mode=Print, stderr_mode=Print) did not execute successfully.
Created at: src/core/builder.rs:1368:21
Executed at: src/core/build_steps/doc.rs:747:22

Build completed unsuccessfully in 1:03:37
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9972818903/job/27559651674)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9972818903/job/27559651674","builder_name":"checks-actions"} -->

tgross35 committed:
Add `core` functions for `f16` and `f128` that require math routines

`min`, `max`, and similar functions require external math routines. Add
these under the same gates as `std` math functions (`reliable_f16_math`
and `reliable_f128_math`).

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit 6a99a101278433ab24d6a242505d11541047ba3b with merge d7fae28da1acfba622f4e27bbdd5cfe98c4ca793...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6a99a101278433ab24d6a242505d11541047ba3b","merge_sha":"d7fae28da1acfba622f4e27bbdd5cfe98c4ca793"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9981257751/job/27587773906)
Build commit: d7fae28da1acfba622f4e27bbdd5cfe98c4ca793 (`d7fae28da1acfba622f4e27bbdd5cfe98c4ca793`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9981257751/job/27587773906"},"merge_sha":"d7fae28da1acfba622f4e27bbdd5cfe98c4ca793"} -->

dtolnay commented on library/core/src/num/f16.rs:
```diff
@@ -739,6 +739,182 @@ impl f16 {
+            // Not safe to halve b
+            (a / 2.) + b
+        } else {
+            // Not safe to halve a and b
```
I suspect this should say "Safe to halve a and b", right?

But I see this is copied from the existing f64::midpoint and f32::midpoint.

tgross35 replied:
I questioned this too 😄 I'll update these, in addition to `f64` and `f32`

tgross35 replied:
Included https://github.com/rust-lang/rust/pull/127027/commits/f92287ac878a7f0db24a161b1a549e938a3b71b1

dtolnay reviewed (approved):
Amazing work!

dtolnay commented:
This is still waiting on an update of the bootstrap compiler to pull in some clippy fix, right? Please use `r=dtolnay` whenever it's ready.

@rustbot blocked

tgross35 commented:
Thank you for the review! Yes, the Clippy ICE fix should be be in beta Friday/Saturday so that should (hopefully) fix CI.

tgross35 committed:
Update comments for `{f16, f32, f64, f128}::midpoint`

Clarify what makes some operations not safe, and correct comment in the
default branch ("not safe" -> "safe").

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 committed:
Add a disclaimer about x86 `f128` math functions

Due to a LLVM bug, `f128` math functions link successfully but LLVM
chooses the wrong symbols (`long double` symbols rather than those for
binary128).

Since this is a notable problem that may surprise a number of users, add
a note about it.

Link: https://github.com/llvm/llvm-project/issues/44744

tgross35 commented:
The stage0 bump took longer than expected but finally merged, so I was able to rebase. Since the last review I removed all the `cfg(not(bootstrap))` and added a note in the docs about the x86 math bug, so users are aware of it.

One more try job to make sure nothing got broken:

@bors try

bors commented:
:hourglass: Trying commit d64bbb127ac920d6058d8f308d0c6670ff07dac7 with merge 2318850ceae2bd34ba063b687e23df424e033e13...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d64bbb127ac920d6058d8f308d0c6670ff07dac7","merge_sha":"2318850ceae2bd34ba063b687e23df424e033e13"} -->

tgross35 commented:
x86 CI passed and it looks like the aarch64 try job has tested the libraries already, so this should be about ready. Changes were trivial since the last look so

@bors r=dtolnay

bors commented:
:pushpin: Commit d64bbb127ac920d6058d8f308d0c6670ff07dac7 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay d64bbb127ac920d6058d8f308d0c6670ff07dac7 -->
<!-- homu: {"type":"Approved","sha":"d64bbb127ac920d6058d8f308d0c6670ff07dac7","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10170681029/job/28132641262)
Approved by: dtolnay
Pushing 2318850ceae2bd34ba063b687e23df424e033e13 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10170681029/job/28132641262"},"merge_sha":"2318850ceae2bd34ba063b687e23df424e033e13"} -->

tgross35 commented:
What the heck
